gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2

```diff
@@ -146,57 +147,47 @@ class {{ service.name }}RestTransport({{ service.name }}Transport):
+        {%- endfor %}
+        },
+        {%- endfor %}]
+
+        request_kwargs = {
+            field.name:value
+            for field, value
+            in {{ method.input.ident }}.pb(request).ListFields()
+        }
+        transcoded_request = path_template.transcode(http_options, **request_kwargs)
+        {%- if 'body' in method.http_options[0] %}
+
         # Jsonify the request body
-        {%- if method.http_opt['body'] != '*' %}
-        body = {{ method.input.fields[method.http_opt['body']].type.ident }}.to_json(
-            request.{{ method.http_opt['body'] }},
+        body = {% if method.http_options[0]['body'] == '*' -%}
+            {{ method.input.ident }}.to_json(
+            {{ method.input.ident }}(transcoded_request['body']),
+        {%- else -%}
+            {{ method.input.fields[method.http_opt['body']].type.ident }}.to_json(
+            {{ method.input.fields[method.http_opt['body']].type.ident }}(transcoded_request['body']),
+        {%- endif %}
             including_default_value_fields=False,
             use_integers_for_enums=False
         )
-        {%- else %}
-        body = {{ method.input.ident }}.to_json(
-            request,
-            use_integers_for_enums=False
-        )
-        {%- endif %}
         {%- endif %}
 
-        {# TODO(yon-mg): Write helper method for handling grpc transcoding url #}
-        # TODO(yon-mg): need to handle grpc transcoding and parse url correctly
-        #               current impl assumes basic case of grpc transcoding
-        url = 'https://{host}{{ method.http_opt['url'] }}'.format(
-            host=self._host,
-            {%- for field in method.path_params %}
-            {{ field }}=request.{{ method.input.get_field(field).name }},
-            {%- endfor %}
-        )
-
-        {# TODO(yon-mg): move all query param logic out of wrappers into here to handle
-                         nested fields correctly (can't just use set of top level fields
-        #}
-        # TODO(yon-mg): handle nested fields corerctly rather than using only top level fields
-        #               not required for GCE
-        query_params = {
-            {%- for field in method.query_params | sort%}
-            '{{ field|camel_case }}': request.{{ field }},
-            {%- endfor %}
-        }
-        # TODO(yon-mg): further discussion needed whether 'python truthiness' is appropriate here
-        #               discards default values
-        # TODO(yon-mg): add test for proper url encoded strings
-        query_params = ['{k}={v}'.format(k=k, v=v) for k, v in query_params.items() if v]
-        url += '?{}'.format('&'.join(query_params)).replace(' ', '+')
+        # Jsonify the query params
+        query_params = json.loads({{ method.input.ident }}.to_json(
+            {{ method.input.ident }}(transcoded_request['query_params']),
+            including_default_value_fields=False,
+            use_integers_for_enums=False
+        ))
```

Comment:
Unless I'm misunderstanding something, this is turning the transcoded request back into a Request message object, turning it into a JSON string, and then turning it back into a python dict. Is that correct? Can we use `to_dict` instead of the indirection?
If the above is correct, can you add some comments describing the dataflow? If it's not correct, then could you describe what the dataflow is?

Reply:
for the record:  this is actually not redundant, provided we want to rely on proto+ to do the proper case conversion.  to_dict doesn't do the conversion from snake_case to camelCase, only to_json does that.  But the result of to_json is a string that can't be used as-is to construct the query string.